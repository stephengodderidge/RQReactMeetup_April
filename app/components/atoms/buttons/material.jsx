import React, { SFC } from 'react';
import Button from '@material-ui/core/Button';
import Fab from '@material-ui/core/Fab';
import { IBaseButtonProps } from '.';

// #region Styling
export const ButtonColor = {
  BLUE: 'primary',
  RED: 'secondary',
  DEFAULT: 'default',
};

export const ButtonSize = {
  SMALL: 'small',
  MEDIUM: 'medium',
  LARGE: 'large',
};
// #endregion Styling

const BaseButton = props => {
  const { id, color, variant, ...otherProps } = props;
  const handleClick = () => props.onClick(id || null);
  const isFab = ['fab', 'extendedFab'].includes(variant);
  const fabVariant = isFab && variant === 'fab' ? 'round' : 'extended';

  return isFab ? (
    <Fab {...otherProps} color={color} onClick={handleClick} variant={fabVariant} />
  ) : (
    <Button {...otherProps} color={color} onClick={handleClick} variant={variant} />
  );
};

export const PrimaryButton = props => <BaseButton {...props} variant="contained" />;

export const SecondaryButton = props => <BaseButton {...props} variant="outlined" />;

export const TertiaryButton = props => <BaseButton {...props} variant="text" />;

export const ActionButton = ({ extended, ...props }) => (
  <BaseButton {...props} variant={extended ? 'extendedFab' : 'fab'} />
);
