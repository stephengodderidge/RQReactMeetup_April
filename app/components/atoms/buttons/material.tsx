import React, { SFC } from 'react';
import Button from '@material-ui/core/Button';
import Fab from '@material-ui/core/Fab';
import { IBaseButtonProps } from '.';
import { withTheme } from '@material-ui/core';

// #region Styling
export enum ButtonColor {
  BLUE = 'primary',
  RED = 'secondary',
  DEFAULT = 'default',
}

export enum ButtonSize {
  SMALL = 'small',
  MEDIUM = 'medium',
  LARGE = 'large',
}

// #endregion Styling

interface IMaterialButtonColorProp {
  /**
   * [optional] Color of button
   *
   * @default GREY
   */
  color?: ButtonColor;
}

interface IMaterialButtonSizeProps {
  /**
   * [optional] Size of button
   *
   * @default MEDIUM
   */
  size?: ButtonSize;
}

type TMaterialButtonVariantProps = IBaseButtonProps &
  IMaterialButtonColorProp &
  IMaterialButtonSizeProps;

export interface IMaterialBaseButtonProps extends TMaterialButtonVariantProps {
  /**
   * [required] Variant of button.  Used to configure button based on 3rd
   * party component library
   *
   * In this app, these values equate to:
   *    contained = PrimaryButton
   *    outlined = SecondaryButton
   *    text = TertiaryButton
   *    fab = ActionButton
   *    extendedFab = ActionButton w/ `extended={true}`
   */
  variant: TMaterialUiVariant;
}

type TMaterialUiVariant = 'text' | 'outlined' | 'contained' | 'fab' | 'extendedFab';

const BaseButton: SFC<IMaterialBaseButtonProps> = props => {
  const { id, color, variant, ...otherProps } = props;
  const handleClick = () => props.onClick(id || null);
  const isFab = ['fab', 'extendedFab'].includes(variant);
  const fabVariant = isFab && variant === 'fab' ? 'round' : 'extended';

  return isFab ? (
    <Fab {...otherProps} color={color} onClick={handleClick} variant={fabVariant} />
  ) : (
    <Button {...otherProps} color={color} onClick={handleClick} variant={variant} />
  );
};

export const PrimaryButton: SFC<TMaterialButtonVariantProps> = props => (
  <BaseButton {...props} variant="contained" />
);

export const SecondaryButton: SFC<TMaterialButtonVariantProps> = props => (
  <BaseButton {...props} variant="outlined" />
);

export const TertiaryButton: SFC<TMaterialButtonVariantProps> = props => (
  <BaseButton {...props} variant="text" />
);

export interface IActionButtonProps
  extends IBaseButtonProps,
    IMaterialButtonColorProp {
  /**
   * [optional] Enables extended Action Button styles for oblong
   * FAB's
   *
   * @default false
   */
  extended?: boolean;
}

export const ActionButton: SFC<IActionButtonProps> = ({ extended, ...props }) => (
  <BaseButton {...props} variant={extended ? 'extendedFab' : 'fab'} />
);

export type TMaterialButtonProps = IMaterialBaseButtonProps | IActionButtonProps;
