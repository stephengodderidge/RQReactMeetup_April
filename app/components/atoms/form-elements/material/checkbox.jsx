import React, { SFC, ChangeEvent } from 'react';
import MaterialCheckbox from '@material-ui/core/Checkbox';
import { MaterialFormElementColors, IMaterialFormToggleElementProps } from '.';

export const Checkbox = props => {
  const onChange = (e, checked) => props.onChange(e.target.value, checked);
  return (
    <MaterialCheckbox
      {...props}
      color={props.color || MaterialFormElementColors.BLUE}
      onChange={onChange}
    />
  );
};
