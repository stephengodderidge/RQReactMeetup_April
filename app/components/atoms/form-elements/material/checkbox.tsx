import React, { SFC, ChangeEvent } from 'react';
import MaterialCheckbox from '@material-ui/core/Checkbox';
import { ISharedMaterialFormElementProps, MaterialFormElementColors } from '.';

export const Checkbox: SFC<ISharedMaterialFormElementProps> = props => {
  const onChange = (e: ChangeEvent<HTMLInputElement>, checked: boolean) =>
    props.onChange(e.target.value, checked);
  return (
    <MaterialCheckbox
      {...props}
      color={props.color || MaterialFormElementColors.BLUE}
      onChange={onChange}
    />
  );
};
