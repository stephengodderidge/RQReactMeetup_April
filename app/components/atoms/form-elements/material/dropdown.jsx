import React, { SFC, ChangeEvent } from 'react';
import { FormControl, Select, MenuItem, InputLabel } from '@material-ui/core/';
import { ISharedMaterialFormElementProps, MaterialFormElementColors } from '.';

export const Dropdown = ({ label, options, ...props }) => {
  const onChange = e => props.onChange(e.target.value);
  return (
    <FormControl className={props.className}>
      <InputLabel htmlFor={label}>{label}</InputLabel>
      <Select
        {...props}
        color={props.color || MaterialFormElementColors.BLUE}
        onChange={onChange}
        inputProps={{
          id: label,
        }}
      >
        {options.map(o => (
          <MenuItem key={o} value={o}>
            {o}
          </MenuItem>
        ))}
      </Select>
    </FormControl>
  );
};
