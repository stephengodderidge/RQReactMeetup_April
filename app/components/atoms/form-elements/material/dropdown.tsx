import React, { SFC, ChangeEvent } from 'react';
import { FormControl, Select, MenuItem, InputLabel } from '@material-ui/core/';
import { ISharedMaterialFormElementProps, MaterialFormElementColors } from '.';

export interface IMaterialFormSelectElementProps
  extends ISharedMaterialFormElementProps {
  /**
   * [required] Callback for when user updates form element
   */
  onChange: (value: string) => void;
  /**
   * [optional] Enables error styling on select element
   */
  error?: boolean;
  /**
   * [required] Options to display in dropdown menu
   */
  options: string[];
  /**
   * [required] Label associated with dropdown
   */
  label: string;
}

export const Dropdown: SFC<IMaterialFormSelectElementProps> = ({
  label,
  options,
  ...props
}) => {
  const onChange = (e: ChangeEvent<HTMLSelectElement>) =>
    props.onChange(e.target.value);
  return (
    <FormControl className={props.className}>
      <InputLabel htmlFor={label}>{label}</InputLabel>
      <Select
        {...props}
        color={props.color || MaterialFormElementColors.BLUE}
        onChange={onChange}
        inputProps={{
          id: label,
        }}
      >
        {options.map(o => (
          <MenuItem key={o} value={o}>
            {o}
          </MenuItem>
        ))}
      </Select>
    </FormControl>
  );
};
