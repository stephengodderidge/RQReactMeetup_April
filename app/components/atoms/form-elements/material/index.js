export * from './checkbox';
export * from './radio-button';
export * from './text-input';
export * from './dropdown';

export const MaterialFormElementColors = {
  BLUE: 'primary',
  RED: 'secondary',
  DEFAULT: 'default',
};
