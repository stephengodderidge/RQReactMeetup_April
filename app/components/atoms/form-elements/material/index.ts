export * from './checkbox';
export * from './radio-button';

export enum MaterialFormElementColors {
  BLUE = 'primary',
  RED = 'secondary',
  DEFAULT = 'default',
}

export interface ISharedMaterialFormElementProps {
  /**
   * [optional] Controls checked state of checkbox
   *
   * @default `false`
   */
  checked?: boolean;
  /**
   * [optional] Checkbox color
   *
   * @default BLUE
   */
  color?: MaterialFormElementColors;
  /**
   * [optional] Enables indeterminate state for checkbox (partial selection)
   */
  indeterminate?: boolean;
  /**
   * [required] Callback for when user interacts with checkbox
   */
  onChange: (value: string, checked: boolean) => void;
  /**
   * [required] Unique value to identify checkbox with.  Will be passed to `onChange` callback
   */
  value: string;
  /**
   * [optional] Disables checkbox interactivity
   */
  disabled?: boolean;
}
