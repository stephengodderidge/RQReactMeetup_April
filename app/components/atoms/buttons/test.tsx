import React from 'react';
import { render, fireEvent } from 'react-testing-library';
import * as MaterialButtons from './material';
import * as SemanticButtons from './semantic';
import * as ReactstrapButtons from './reactstrap';

const onClick = jest.fn();

describe('Buttons', () => {
  afterEach(() => {
    onClick.mockClear();
  });
  // #region Material UI Button Tests
  describe('Material', () => {
    describe('Primary Button', () => {
      test('Handles onClick', () => {
        const testId = 'material-primary';
        const { getByTestId } = render(
          <MaterialButtons.PrimaryButton data-testid={testId} onClick={onClick}>
            Clickable Button
          </MaterialButtons.PrimaryButton>,
        );
        fireEvent.click(getByTestId(testId));
        expect(onClick).toHaveBeenCalledTimes(1);
      });
      test('Renders Primary Style', () => {
        const { container } = render(
          <MaterialButtons.PrimaryButton onClick={onClick}>
            Primary
          </MaterialButtons.PrimaryButton>,
        );
        expect(container).toMatchSnapshot();
      });
      test('Renders Primary Style - BLUE', () => {
        const { container } = render(
          <MaterialButtons.PrimaryButton
            onClick={onClick}
            color={MaterialButtons.ButtonColor.BLUE}
          >
            Primary
          </MaterialButtons.PrimaryButton>,
        );
        expect(container).toMatchSnapshot();
      });
      test('Renders Secondary Style', () => {
        const { container } = render(
          <MaterialButtons.SecondaryButton onClick={onClick}>
            Secondary
          </MaterialButtons.SecondaryButton>,
        );
        expect(container).toMatchSnapshot();
      });
      test('Renders Secondary Style - BLUE', () => {
        const { container } = render(
          <MaterialButtons.SecondaryButton
            onClick={onClick}
            color={MaterialButtons.ButtonColor.BLUE}
          >
            Secondary
          </MaterialButtons.SecondaryButton>,
        );
        expect(container).toMatchSnapshot();
      });
      test('Renders Tertiary Style', () => {
        const { container } = render(
          <MaterialButtons.TertiaryButton onClick={onClick}>
            Tertiary
          </MaterialButtons.TertiaryButton>,
        );
        expect(container).toMatchSnapshot();
      });
      test('Renders Action Style', () => {
        const { container } = render(
          <MaterialButtons.ActionButton onClick={onClick}>
            Action
          </MaterialButtons.ActionButton>,
        );
        expect(container).toMatchSnapshot();
      });
      test('Renders Action Style - BLUE', () => {
        const { container } = render(
          <MaterialButtons.ActionButton
            onClick={onClick}
            color={MaterialButtons.ButtonColor.BLUE}
          >
            Action
          </MaterialButtons.ActionButton>,
        );
        expect(container).toMatchSnapshot();
      });
      test('Renders Extended Action Style', () => {
        const { container } = render(
          <MaterialButtons.ActionButton extended onClick={onClick}>
            Extended Action
          </MaterialButtons.ActionButton>,
        );
        expect(container).toMatchSnapshot();
      });
    });
  });
  // #endregion Material UI Button Tests
  // #region Semantic UI Button Tests
  describe('Semantic', () => {
    describe('Primary Button', () => {
      test('Handles onClick', () => {
        const testId = 'semantic-primary';
        const { getByTestId } = render(
          <SemanticButtons.PrimaryButton data-testid={testId} onClick={onClick}>
            Clickable Button
          </SemanticButtons.PrimaryButton>,
        );
        fireEvent.click(getByTestId(testId));
        expect(onClick).toHaveBeenCalledTimes(1);
      });
      test('Renders Primary Style', () => {
        const { container } = render(
          <SemanticButtons.PrimaryButton onClick={onClick}>
            Primary
          </SemanticButtons.PrimaryButton>,
        );
        expect(container).toMatchSnapshot();
      });
      test('Renders Secondary Style', () => {
        const { container } = render(
          <SemanticButtons.SecondaryButton onClick={onClick}>
            Secondary
          </SemanticButtons.SecondaryButton>,
        );
        expect(container).toMatchSnapshot();
      });
      test('Renders Tertiary Style', () => {
        const { container } = render(
          <SemanticButtons.TertiaryButton onClick={onClick}>
            Tertiary
          </SemanticButtons.TertiaryButton>,
        );
        expect(container).toMatchSnapshot();
      });
    });
  });
  // #endregion Semantic UI Button Tests
  // #region Reactstrap UI Button Tests
  describe('Reactstrap', () => {
    describe('Primary Button', () => {
      test('Handles onClick', () => {
        const testId = 'reactstrap-primary';
        const { getByTestId } = render(
          <ReactstrapButtons.PrimaryButton data-testid={testId} onClick={onClick}>
            Clickable Button
          </ReactstrapButtons.PrimaryButton>,
        );
        fireEvent.click(getByTestId(testId));
        expect(onClick).toHaveBeenCalledTimes(1);
      });
      test('Renders Primary Style', () => {
        const { container } = render(
          <ReactstrapButtons.PrimaryButton onClick={onClick}>
            Primary
          </ReactstrapButtons.PrimaryButton>,
        );
        expect(container).toMatchSnapshot();
      });
      test('Renders Secondary Style', () => {
        const { container } = render(
          <ReactstrapButtons.SecondaryButton onClick={onClick}>
            Secondary
          </ReactstrapButtons.SecondaryButton>,
        );
        expect(container).toMatchSnapshot();
      });
    });
  });
  // #endregion Reactstrap UI Button Tests
});
