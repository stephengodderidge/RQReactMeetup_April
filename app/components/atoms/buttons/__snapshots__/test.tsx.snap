// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Buttons Material Primary Button Renders Action Style - GREEN 1`] = `
<div>
  <button
    class="MuiButtonBase-root-27 MuiFab-root-37"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiFab-label-38"
    >
      Action
    </span>
    <span
      class="MuiTouchRipple-root-30"
    />
  </button>
</div>
`;

exports[`Buttons Material Primary Button Renders Action Style 1`] = `
<div>
  <button
    class="MuiButtonBase-root-27 MuiFab-root-37"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiFab-label-38"
    >
      Action
    </span>
    <span
      class="MuiTouchRipple-root-30"
    />
  </button>
</div>
`;

exports[`Buttons Material Primary Button Renders Extended Action Style 1`] = `
<div>
  <button
    class="MuiButtonBase-root-27 MuiFab-root-37 MuiFab-extended-41"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiFab-label-38"
    >
      Extended Action
    </span>
    <span
      class="MuiTouchRipple-root-30"
    />
  </button>
</div>
`;

exports[`Buttons Material Primary Button Renders Primary Style - GREEN 1`] = `
<div>
  <button
    class="MuiButtonBase-root-27 MuiButton-root-1 MuiButton-contained-12 MuiButton-raised-15"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label-2"
    >
      Primary
    </span>
    <span
      class="MuiTouchRipple-root-30"
    />
  </button>
</div>
`;

exports[`Buttons Material Primary Button Renders Primary Style 1`] = `
<div>
  <button
    class="MuiButtonBase-root-27 MuiButton-root-1 MuiButton-contained-12 MuiButton-raised-15"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label-2"
    >
      Primary
    </span>
    <span
      class="MuiTouchRipple-root-30"
    />
  </button>
</div>
`;

exports[`Buttons Material Primary Button Renders Secondary Style - GREEN 1`] = `
<div>
  <button
    class="MuiButtonBase-root-27 MuiButton-root-1 MuiButton-outlined-9"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label-2"
    >
      Secondary
    </span>
    <span
      class="MuiTouchRipple-root-30"
    />
  </button>
</div>
`;

exports[`Buttons Material Primary Button Renders Secondary Style 1`] = `
<div>
  <button
    class="MuiButtonBase-root-27 MuiButton-root-1 MuiButton-outlined-9"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label-2"
    >
      Secondary
    </span>
    <span
      class="MuiTouchRipple-root-30"
    />
  </button>
</div>
`;

exports[`Buttons Material Primary Button Renders Tertiary Style 1`] = `
<div>
  <button
    class="MuiButtonBase-root-27 MuiButton-root-1 MuiButton-text-3 MuiButton-flat-6"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label-2"
    >
      Tertiary
    </span>
    <span
      class="MuiTouchRipple-root-30"
    />
  </button>
</div>
`;

exports[`Buttons Reactstrap Primary Button Renders Primary Style 1`] = `
<div>
  <button
    class="btn btn-secondary"
    type="button"
  >
    Primary
  </button>
</div>
`;

exports[`Buttons Reactstrap Primary Button Renders Secondary Style 1`] = `
<div>
  <button
    class="btn btn-outline-secondary"
    type="button"
  >
    Secondary
  </button>
</div>
`;

exports[`Buttons Semantic Primary Button Renders Primary Style 1`] = `
<div>
  <button
    class="ui button"
  >
    Primary
  </button>
</div>
`;

exports[`Buttons Semantic Primary Button Renders Secondary Style 1`] = `
<div>
  <button
    class="ui basic button"
  >
    Secondary
  </button>
</div>
`;

exports[`Buttons Semantic Primary Button Renders Tertiary Style 1`] = `
.c0 .ui.basic.button {
  box-shadow: none !important;
}

<div>
  <div
    class="c0"
  >
    <button
      class="ui basic button"
    >
      Tertiary
    </button>
  </div>
</div>
`;
